src/HDF5.jl

```diff
@@ -302,7 +302,21 @@ struct Opaque
 end
 
 # An empty array type
-struct EmptyArray{T} end
+struct EmptyArray{T} <: AbstractArray{T,0} end
+# Required AbstractArray interface
+Base.size(::EmptyArray) = ()
+Base.IndexStyle(::Type{<:EmptyArray}) = IndexLinear()
+Base.getindex(::EmptyArray, ::Int) = error("cannot index an `EmptyArray`")
+Base.setindex!(::EmptyArray, v, ::Int) = error("cannot assign to an `EmptyArray`")
+# Optional interface
+Base.similar(::EmptyArray{T}) where {T} = EmptyArray{T}()
+Base.similar(::EmptyArray, ::Type{S}) where {S} = EmptyArray{S}()
+Base.similar(::EmptyArray, ::Type{S}, dims::Dims) where {S} = Array{S}(undef, dims)
+# Override behavior for 0-dimensional Array
+Base.length(::EmptyArray) = 0
+# Required to avoid indexing during printing
+Base.show(io::IO, E::EmptyArray) = print(io, typeof(E), "()")
+Base.show(io::IO, ::MIME"text/plain", E::EmptyArray) = show(io, E)
```

Comment:
Isn't this already the default? 

From docs:

>   The default MIME type is MIME"text/plain". There is a fallback definition for text/plain output that calls show with 2 arguments, so it is not always
  necessary to add a method for that case.

Reply:
There's a default implementation of 3-arg show for `<:AbstractArray`, so that has to be overridden — https://github.com/JuliaLang/julia/blob/44d44708f0f0f2456c623bcd999a89adc35fe8c8/base/arrayshow.jl#L323

Reply:
Oh right that makes sense. I guess we are still missing something w.r.t to printing

The matrix case ends up here printing junk values
![image](https://user-images.githubusercontent.com/4319522/96151846-e05fd100-0ed9-11eb-9f6e-d654eff558e2.png)

Reply:
The last one is `Matrix{Int32}`, so something weird went on with `hvcat`, I think?

Reply:
Hmm, interesting, didn't even realize the type got changes.

Reply:
I think this is because Base's 0-dimensional arrays have shape `()` but have a size of `1` in any dimension when being concatenated with other arrays. We actually want an empty shape to have zero size, so there's a conflict with Base's assumption, and it's "overallocating" the output array.

This seems to be buried pretty deep in Base, so I'm not sure if there's a good fix on this end — various hacks I've tried end up exposing different unwanted behavior. The best option might be to just forceably error out within the concatenation path? I think an overload of `Base.cat_size(::HDF5.EmptyArray) = error("...unsupported...")` would do it.

Reply:
Hmm that seems annoying. I guess we can either just leave it or add the  overload you suggest iable option (although this might end up being wack-a-mole, if we discover other similar methods.).

Reply:
The other option is to drop the `<: AbstractArray` inheritance, which then makes the concatenation framework treat the type as a scalar. The problem there is that again you have to play whack-a-mole defining a bunch of other methods which are assumed for any array-like type (and a bunch of methods — `write`, `d_write`, etc) have to have their types modified to accept a non-`AbstractArray` type explicitly... No good options.

Reply:
Yeah that sounds like the worse option.